File: 933. 最近的请求次数/main.ts
const RANGE = 3000;
class RecentCounter {
    private queue: number[];
    constructor() {
        this.queue = []
    }

    ping(t: number): number {
        while (this.queue.length > 0 && this.queue[0] < t - RANGE) {
            this.queue.shift();
        }
        this.queue.push(t);
        return this.queue.length;
    }
}

/**
 * Your RecentCounter object will be instantiated and called as such:
 * var obj = new RecentCounter()
 * var param_1 = obj.ping(t)
 */

File: 933. 最近的请求次数/main.rs
use std::collections::VecDeque;

struct RecentCounter {
    q: VecDeque<i32>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl RecentCounter {
    fn new() -> Self {
        Self { q: VecDeque::new() }
    }

    fn ping(&mut self, t: i32) -> i32 {
        self.q.push_back(t);
        while let Some(&front) = self.q.front() {
            if front < t - 3000 {
                self.q.pop_front();
            } else {
                break;
            }
        }
        self.q.len() as i32
    }
}/**
 * Your RecentCounter object will be instantiated and called as such:
 * let obj = RecentCounter::new();
 * let ret_1: i32 = obj.ping(t);
 */
